otherlibs/systhreads/st_win32.h

```diff
@@ -24,7 +24,242 @@ Caml_inline void st_msleep(int msec)
+  return TlsGetValue(k);
+}
+
+Caml_inline void st_tls_set(st_tlskey k, void * v)
+{
+  (void)TlsSetValue(k, v);
+}
+
+/* The master lock.  This is a mutex that is held most of the time,
+   so we implement it in a slightly convoluted way to avoid
+   all risks of busy-waiting.  Also, we count the number of waiting
+   threads. */
+
+typedef struct {
+  bool init;                      /* have the mutex and the cond been
+                                     initialized already? */
+  SRWLOCK lock;                   /* to protect contents */
+  bool busy;                      /* false = free, true = taken */
+  atomic_uintnat waiters;         /* number of threads waiting on master lock */
+  CONDITION_VARIABLE is_free;     /* signaled when free */
+} st_masterlock;
+
+/* Returns non-zero on failure */
+static int st_masterlock_init(st_masterlock * m)
+{
+  if (!m->init) {
+    InitializeSRWLock(&m->lock);
+    InitializeConditionVariable(&m->is_free);
+    m->init = true;
+  }
+  m->busy = true;
+  atomic_store_release(&m->waiters, 0);
+  return 0;
+}
+
+static uintnat st_masterlock_waiters(st_masterlock * m)
+{
+  return atomic_load_acquire(&m->waiters);
+}
+
+static void st_masterlock_acquire(st_masterlock *m)
+{
+  AcquireSRWLockExclusive(&m->lock);
+  while (m->busy) {
+    atomic_fetch_add(&m->waiters, +1);
+    SleepConditionVariableSRW(&m->is_free, &m->lock,
+                              INFINITE, 0 /* exclusive */);
+    atomic_fetch_add(&m->waiters, -1);
+  }
+  m->busy = true;
+  st_bt_lock_acquire();
+  ReleaseSRWLockExclusive(&m->lock);
+
+  return;
+}
+
+static void st_masterlock_release(st_masterlock * m)
+{
+  AcquireSRWLockExclusive(&m->lock);
+  m->busy = false;
+  st_bt_lock_release(st_masterlock_waiters(m) == 0);
+  WakeConditionVariable(&m->is_free);
+  ReleaseSRWLockExclusive(&m->lock);
+
+  return;
+}
+
+/* Scheduling hints */
+
+/* This is mostly equivalent to release(); acquire(), but better. In particular,
+   release(); acquire(); leaves both us and the waiter we signal() racing to
+   acquire the lock. Calling yield or sleep helps there but does not solve the
+   problem. Sleeping ourselves is much more reliable--and since we're handing
+   off the lock to a waiter we know exists, it's safe, as they'll certainly
+   re-wake us later.
+*/
+Caml_inline void st_thread_yield(st_masterlock * m)
+{
+  AcquireSRWLockExclusive(&m->lock);
+  /* We must hold the lock to call this. */
+
+  /* We already checked this without the lock, but we might have raced--if
+     there's no waiter, there's nothing to do and no one to wake us if we did
+     wait, so just keep going. */
+  uintnat waiters = st_masterlock_waiters(m);
+
+  if (waiters == 0) {
+    ReleaseSRWLockExclusive(&m->lock);
+    return;
+  }
+
+  m->busy = false;
+  atomic_fetch_add(&m->waiters, +1);
+  WakeConditionVariable(&m->is_free);
+  /* releasing the domain lock but not triggering bt messaging
+     messaging the bt should not be required because yield assumes
```

Comment:
```suggestion
  /* releasing the domain lock but not triggering bt messaging.
     Messaging the bt should not be required because yield assumes
```